src/Nancy.Hosting.Self/NancyHost.cs

```diff
@@ -124,20 +125,27 @@ public void Start()
         {
             this.StartListener();
 
-            Task.Factory.StartNew(async () =>
+            Task.Run(() =>
             {
-                try
+                var semaphore = new Semaphore(this.configuration.MaximumConnectionCount, this.configuration.MaximumConnectionCount);
+                while (!this.stop)
                 {
-                    while(!this.stop)
+                    semaphore.WaitOne();
+
+                    this.listener.GetContextAsync().ContinueWith(async (contextTask) =>
                     {
-                        HttpListenerContext context = await this.listener.GetContextAsync().ConfigureAwait(false);
-                        await this.Process(context).ConfigureAwait(false);
-                    }
-                }
-                catch(Exception ex)
-                {
-                    this.configuration.UnhandledExceptionCallback.Invoke(ex);
-                    throw;
+                        try
+                        {
+                            semaphore.Release();

```

Comment:
Shouldn't the `Release()` be in a `finally` clause? If the above `GetcontextAsyn().ContinueWith()` statement throws, afaict, the lock will never be released. Whether you actually want a deadlock or not is a difficult choice to make, but [Eric Lippert outlines the different options in this blog post](https://blogs.msdn.microsoft.com/ericlippert/2009/03/06/locks-and-exceptions-do-not-mix/).